breathe/directives.py

```diff
@@ -138,40 +138,21 @@ def run(self) -> List[Node]:
                            self.directive_args)
 
     def parse_args(self, function_description):
-        # Strip off trailing qualifiers
-        pattern = re.compile(r'''(?<= \)) \s*
-                             (?: = \s* 0)? \s* $ ''',
-                             re.VERBOSE)
-
-        function_description = re.sub(pattern,
-                                      '',
-                                      function_description)
-
-        paren_index = function_description.find('(')
-        if paren_index == -1:
-            return []
-        # If it is empty parenthesis, then return empty list as we want empty parenthesis coming
-        # from the xml file to match the user's function when the user doesn't provide parenthesis
-        # ie. when there are no args anyway
-        elif function_description == '()':
-            return []
-        else:
-            # Parse the function name string, eg. f(int, float) to
-            # extract the types so we can use them for matching
-            args = []
-            num_open_brackets = -1
-            start = paren_index + 1
-            for i in range(paren_index, len(function_description)):
-                c = function_description[i]
-                if c == '(' or c == '<':
-                    num_open_brackets += 1
-                elif c == ')' or c == '>':
-                    num_open_brackets -= 1
-                elif c == ',' and num_open_brackets == 0:
-                    args.append(function_description[start:i].strip())
-                    start = i + 1
-            args.append(function_description[start:-1].strip())
-            return args
+        if function_description == '':
+            function_description = '()'
+
+        parser = cpp.DefinitionParser(function_description,
+                                      location=self.get_source_info(),
+                                      config=self.config)
```

Comment:
All the projects I am working on are in C, but would this work if I decided to use `.. doxygenfunction::`?

Reply:
In general, no, but in practice, yes: for C I would expect that the argument you give will always be just a single identifier, without a parameter list, so this code is never executed. The different form, with a parameter list, is for disambiguation of overloads. Though, if you write a C parameter list, it will work anyway, as long as it is valid C++ (e.g., no ``restrict`` qualifier).
There is also the option of writing qualified names for the directive, but this handling is already C++-specific as it assumes ``::`` as scope 'operator'.
Though, all this really should be tested to be sure (perhaps there already is a bit in the doc folder somewhere).

Reply:
I tested it out on a C project I'm working on and can confirm it works properly over there, no issues.

Reply:
Some further testing shows it does call it many times for C, but the `function_description` is always the empty string. I also tested `restrict` somewhat by adding the keyword to an existing header, resulting in something like this.

```c
void func(const char *restrict s)
```

```rst
.. doxygenfunction:: func
```

Which worked fine. If I understand correctly it is problematic if you fully specify the parameter list on the `doxygenfunction` directive. Doing that doesn't make much sense in C, so I'm perfectly fine with not fully supporting that in some edge cases.